gridfs/grid_file.py

```diff
@@ -268,32 +277,47 @@ def __setattr__(self, name: str, value: Any) -> None:
             if self._closed:
                 self._coll.files.update_one({"_id": self._file["_id"]}, {"$set": {name: value}})
 
-    def __flush_data(self, data: Any) -> None:
+    def __flush_data(self, data: Any, force: bool = False) -> None:
         """Flush `data` to a chunk."""
         self.__ensure_indexes()
-        if not data:
-            return
         assert len(data) <= self.chunk_size
-
-        chunk = {"files_id": self._file["_id"], "n": self._chunk_number, "data": Binary(data)}
-
-        try:
-            self._chunks.insert_one(chunk, session=self._session)
-        except DuplicateKeyError:
-            self._raise_file_exists(self._file["_id"])
+        if data:
+            self._buffered_docs.append(
+                {"files_id": self._file["_id"], "n": self._chunk_number, "data": Binary(data)}
+            )
+            self._buffered_docs_size += len(data)
```

Comment:
Should this include rest of the document size as well?

Reply:
Hmm good point. This won't matter much in practice because the default chunk_size (256KB) dominates the document overhead but I could see some pathological cases where the app uses chunk_size=1 or a very large "files_id" where this could behave poorly. In those cases we'd batch up far more documents in memory than we'd need.

I see two fixes that are simpler than guestimating the document size:
- encode to RawBSONDocument so we know the real size
- limit the batch to at most 100,000 documents.

Reply:
I decided not to use RawBSONDocument as is was tricky to get right (which codec options to use), is less readable code-wise, and is a little less efficient (due to extra memory copies shuffling around the raw documents). 

I did add the 100,000 limit (OP_MSG's maxWriteBatchSize) and added some extra overhead to account for the size of the chunk doc encoded as bson.